### PR TITLE
Add refresh logic for REST objects

### PR DESCRIPTION
**Pull request by ibmkendrick**
* override `__getattr__` to access JSON response attributes without modifying the `_ResourceElement`'s attribute dictionary
* override `__setattr__` to disallow setting of attributes names that are identical with JSON response property names.

Fixes IBMStreams/streamsx.topology#862

### CODE REVIEW

**Comment by ddebrunner** [comment]
Can one of the admins verify this patch?

**Comment by ibmkendrick** [comment]
@ddebrunner, @wmarshall484 .  Figured out an alternate way to get the JSON response properties in `_ResourceElement` without adding the properties directly into the object.

It still doesn't handle REST errors (as we have discussed earlier), but it does offer the refresh() functionality without breaking the existing design.

I will add some unit tests once #869 is merged.  (As the unit test will require StreamsConnection)

**Comment by ddebrunner** [comment]
test now

**Comment by wmarshall484** [comment]
+1 on overriding `__getattr__` and `__setattr__`, it doesn't pollute the Python object's `__dict__`.

**Comment by ddebrunner** [comment]
Would need some tests before merging this PR.

**Comment by ibmkendrick** [comment]
Waiting for #870 to be merged, before adding unit test.  (As it requires the new test infrastructure)

**Comment by ddebrunner** [comment]
@wmarshall484 Can you approve(?) the ode review?

**Comment by wmarshall484** [comment]
Can this be merged?

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py [hunk 1]
```diff
@@ -37,19 +37,41 @@ class _ResourceElement(object):
...
+        self.rest_client = rest_client
+        self.rest_self = json_rep.get('self', None)
+        self.json_rep = json_rep
 
     def __str__(self):
         return pformat(self.__dict__)
 
+    # Override getattr to retrieve attribute from response JSON
+    def __getattr__(self, key):
+        if 'json_rep' in self.__dict__:
```
Comment (wmarshall484) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py]: Ok, now I'm confused. On this line when you call `self.__dict__`, why doesn't that recurse forever and cause a stack overflow?

Comment (wmarshall484) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py]: Since `self.__dict__` will go and call `__getattr__` again.

Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py]: I think `__getattr__` is only called if the attribute is not found in the object.   So `self.__dict__` will call `__getattribute__`, but not `__getattr__`

**Comment by ibmkendrick** [comment]
@ddebrunner , unit test added: eb91df1

File: com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py (hunk 1 above)
Comment (ibmkendrick) [inline com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py]:
```
class myobj(object):
  def __init__(self):
    self.a = 'a'

  def __getattr__(self, key):
    print ("calling __getattr__ with {}".format(key))
    return super(myobj, self).__getattribute__(key)

  def __getattribute__(self, key):
    print ("calling __getattribute__ with {}".format(key))
    return super(myobj, self).__getattribute__(key)

o = myobj()
o.a
o.__dict__
o.b
```

will yield:
```
calling __getattribute__ with a
calling __getattribute__ with __dict__
calling __getattribute__ with b
calling __getattr__ with b
```

**Comment by ddebrunner** [comment]
test now

**Pull request merged**